include/aws/common/array_list.h

```diff
@@ -0,0 +1,265 @@
+                                                                                                                       \
+            memcpy(raw_data, list->data, list->length * sizeof(T));                                                    \
+            aws_mem_release(list->alloc, (void *)list->data);                                                          \
+            list->data = raw_data;                                                                                     \
+            list->current_size = list->length;                                                                         \
+        }                                                                                                              \
+                                                                                                                       \
+        return AWS_OP_SUCCESS;                                                                                         \
+    }                                                                                                                  \
+                                                                                                                       \
+    return aws_raise_error(AWS_ERROR_LIST_STATIC_MODE_CANT_SHRINK);                                                    \
+}                                                                                                                      \
+                                                                                                                       \
+/**
+* Copies the elements from from to to. to must at least be the same length as from.
+*/                                                                                                                     \
+static inline int aws_##name##_array_list_copy(const struct aws_##name##_array_list *from,                             \
+                        struct aws_##name##_array_list *to) {                                                          \
+    if (to->current_size >= from->length) {                                                                            \
+        memcpy(to->data, from->data, from->length * sizeof(T));                                                        \
+        to->length = from->length;                                                                                     \
+        return AWS_OP_SUCCESS;                                                                                         \
+    }                                                                                                                  \
+                                                                                                                       \
+    return aws_raise_error(AWS_ERROR_LIST_DEST_COPY_TOO_SMALL);                                                        \
```

Comment:
Why doesn't this allocate space for a dynamic list?

Reply:
Fair. Will add.